### PR TITLE
Implement suspendRead and resumeRead in Java

### PR DESCRIPTION
**Pull request by jfullerton44**
### Changes proposed: 
- Implements the suspendRead and ResumeRead functions in WebSocketSession using a semaphore

### CODE REVIEW

File: java/src/main/java/com/microsoft/tunnels/websocket/WebSocketSession.java [hunk 1]
```diff
@@ -81,13 +85,13 @@ public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) throws E
...
-  public void suspendRead()  {
-    super.suspendRead();
+  public void suspendRead() {
+    this.reading.tryAcquire();
   }
 
   @Override
   public void resumeRead()  {
-    super.resumeRead();
+    this.reading.release();
```
Comment (IlyaBiryukov) [inline java/src/main/java/com/microsoft/tunnels/websocket/WebSocketSession.java:94]: What if `suspendRead` couldn't acquire the semaphore?

File: java/src/main/java/com/microsoft/tunnels/websocket/WebSocketSession.java [hunk 2]
```diff
@@ -72,6 +74,8 @@ public void channelInactive(ChannelHandlerContext ctx) throws Exception {
 
   @Override
   public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
+    this.reading.acquire();
+    this.reading.release();
```
Comment (IlyaBiryukov) [inline java/src/main/java/com/microsoft/tunnels/websocket/WebSocketSession.java:78]: Should the release happen after `super.channelRead()`?

File: java/src/main/java/com/microsoft/tunnels/websocket/WebSocketSession.java [hunk 3]
```diff
@@ -20,7 +21,7 @@
 
 public class WebSocketSession extends NettyIoSession {
   protected WebSocketConnectionHandler webSocketConnectionHandler;
-
+  protected Semaphore reading;
```
Comment (IlyaBiryukov) [inline java/src/main/java/com/microsoft/tunnels/websocket/WebSocketSession.java:24]: Can it be `private`?

File: java/src/main/java/com/microsoft/tunnels/websocket/WebSocketSession.java (hunk 1 above)
Comment (jfullerton44) [reply java/src/main/java/com/microsoft/tunnels/websocket/WebSocketSession.java:98]: If it cant acquire the semaphore that means that suspendRead was already called. The tryAcquire will always run before Acquire

File: java/src/main/java/com/microsoft/tunnels/websocket/WebSocketSession.java (hunk 2 above)
Comment (jfullerton44) [reply java/src/main/java/com/microsoft/tunnels/websocket/WebSocketSession.java:78]: Multiple reads are allowed at the same time, we just need to check if reads are blocked before starting

File: java/src/main/java/com/microsoft/tunnels/websocket/WebSocketSession.java (hunk 3 above)
Comment (jfullerton44) [reply java/src/main/java/com/microsoft/tunnels/websocket/WebSocketSession.java]: updated

File: java/src/main/java/com/microsoft/tunnels/websocket/WebSocketSession.java (hunk 1 above)
Comment (jfullerton44) [reply java/src/main/java/com/microsoft/tunnels/websocket/WebSocketSession.java:98]: if the semaphore is available when resumeRead is called it just returns immediately

**Review by IlyaBiryukov — APPROVED** [review]

**Pull request merged**